lib/codecs/src/decoding/format/json_paths.rs

```diff
@@ -0,0 +1,1182 @@
+
+        let input = Bytes::from(
+            r#"{"meta": {"source": "foo"}, "results": {"records": [{"log": "bar"}, {"log": "baz"}]}, "tail": "foo bar baz"}"#,
+        );
+        let events = deserializer
+            .parse(input, LogNamespace::Vector)
+            .unwrap();
+
+        assert_eq!(events.len(), 4);
+
+        let exprs: Vec<Value> = events
+            .iter()
+            .map(|e| e.as_log()["expr"].clone())
+            .collect();
+
+        assert!(exprs.contains(&"meta".into()));
+        assert!(exprs.iter().filter(|e| *e == &Value::from("results.records")).count() == 2);
+        assert!(exprs.contains(&"tail".into()));
+    }
+
+    #[test]
+    fn test_order_preservation() {
+        let config = PathOperationConfig::new(
+            owned_value_path!("meta"),
+            PathOperation::Identity,
+        )
+        .with_path(owned_value_path!("results", "records"), PathOperation::Explode)
+        .with_path(owned_value_path!("tail"), PathOperation::Bytes);
+        let deserializer = JsonPathDeserializer::new(config, JsonPathDeserializerOptions::default());
+
+        let input = Bytes::from(
+            r#"{"meta": {"source": "foo"}, "results": {"records": [{"log": "bar"}, {"log": "baz"}]}, "tail": "foo bar baz"}"#,
+        );
+        let events = deserializer
+            .parse(input, LogNamespace::Vector)
+            .unwrap();
+
+        let exprs: Vec<Value> = events
+            .iter()
+            .map(|e| e.as_log()["expr"].clone())
+            .collect();
+
+        assert_eq!(events.len(), 4);
+        assert_eq!(exprs[0], "meta".into());
+        assert_eq!(exprs[1], "results.records".into());
+        assert_eq!(exprs[2], "results.records".into());
+        assert_eq!(exprs[3], "tail".into());
+    }
+
+    #[test]
+    fn test_multiple_concatenated_json() {
+        let config = PathOperationConfig::new(
+            owned_value_path!("meta"),
+            PathOperation::Identity,
+        )
+        .with_path(owned_value_path!("results", "records"), PathOperation::Explode);
+        let deserializer = JsonPathDeserializer::new(config, JsonPathDeserializerOptions::default());
+
+        let input = Bytes::from(
+            r#"{"meta": {"source": "first"}, "results": {"records": [{"log": "a"}]}}{"meta": {"source": "second"}, "results": {"records": [{"log": "b"}, {"log": "c"}]}}"#,
+        );
+        let events = deserializer
+            .parse(input, LogNamespace::Vector)
+            .unwrap();
+
+        assert_eq!(events.len(), 5);
+
+        let exprs: Vec<Value> = events
+            .iter()
+            .map(|e| e.as_log()["expr"].clone())
+            .collect();
+
+        assert_eq!(exprs[0], "meta".into());
+        assert_eq!(exprs[1], "results.records".into());
+        assert_eq!(exprs[2], "meta".into());
+        assert_eq!(exprs[3], "results.records".into());
+        assert_eq!(exprs[4], "results.records".into());
```

Comment:
ditto, assert on data too? (in all the other tests too), we can only pick a single attribute to assert on perhaps to make it cleaner?